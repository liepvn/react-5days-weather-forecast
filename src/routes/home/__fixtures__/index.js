export * from './home-state.factory';