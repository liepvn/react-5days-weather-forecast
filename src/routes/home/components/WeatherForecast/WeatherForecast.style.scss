@import "~bootstrap/scss/bootstrap";

.weather-forecast {
  .default-zone {
    height: 10em;
    background-color: $gray-200;
    display: flex;
    justify-content: center;
    align-items: center;
  }
}