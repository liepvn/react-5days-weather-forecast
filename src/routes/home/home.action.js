import { createAction } from "core/helpers";

export const ActionTypes = {
  FETCH_WEATHER_FORECAST_REQUEST: 'FETCH_WEATHER_FORECAST_REQUEST',
  FETCH_WEATHER_FORECAST_SUCCESS: 'FETCH_WEATHER_FORECAST_SUCCESS',
  FETCH_WEATHER_FORECAST_FAILURE: 'FETCH_WEATHER_FORECAST_FAILURE',
  SEARCH_LOCATIONS_REQUEST: 'SEARCH_LOCATIONS_REQUEST',
  SEARCH_LOCATIONS_SUCCESS: 'SEARCH_LOCATIONS_SUCCESS',
  SEARCH_LOCATIONS_FAILURE: 'SEARCH_LOCATIONS_FAILURE',
}

export const fetchWeatherForecast = createAction(ActionTypes.FETCH_WEATHER_FORECAST_REQUEST, ({ woeid }) => ({ woeid }));
export const searchLocations = createAction(ActionTypes.SEARCH_LOCATIONS_REQUEST, (query) => ({ query }));