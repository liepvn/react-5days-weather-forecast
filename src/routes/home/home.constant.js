export const WEATHER_STATE_IMAGE_BASE_URL = 'https://www.metaweather.com/static/img/weather/png/64';

export const DEFAULT_LOCATION = Object.freeze({ woeid: 1252431, title: 'Ho Chi Minh City' });