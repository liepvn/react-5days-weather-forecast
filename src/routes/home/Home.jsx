import React from 'react';
import { Col, Container, Row } from 'react-bootstrap';

import { WeatherForecast } from './components/WeatherForecast';
import { DEFAULT_LOCATION } from './home.constant';

/**
 * Home component
 *
 * @returns {JSX.Element}
 */
const Home = () => {
  return (
    <Container>
      <Row>
        <Col lg={{ span: 6, offset: 3 }} className='pt-4'>
          <WeatherForecast defaultLocation={DEFAULT_LOCATION} />
        </Col>
      </Row>
      <Row>
        <a
          href='https://github.com/liepvn'
          target='_blank'
          rel='noopener noreferrer'
          className='text-white text-center text-decoration-none mt-3'
        >
          <i className='bi bi-github'></i> <span>Fork it on Github</span>
        </a>
      </Row>
    </Container>
  );
};

export default Home;
