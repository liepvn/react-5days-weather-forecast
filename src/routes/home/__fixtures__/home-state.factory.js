import { cloneDeep } from 'lodash';

import { homeState } from 'routes/home/home.reducer';

class HomeStateFactory {
  newState = {};

  build(props = {}) {
    this.newState = Object.assign(cloneDeep(homeState), props);

    return this;
  }

  withSampleForecast() {
    Object.assign(this.newState, {
      forecast: [
        {
          maxTemp: 32.06,
          minTemp: 24.880000000000003,
          date: '2021-08-15',
          weatherStateAbbr: 'hr',
          weatherStateImage: 'https://www.metaweather.com/static/img/weather/png/64/hr.png',
          humidity: 73,
          temp: 30.515,
        },
        {
          maxTemp: 33.995000000000005,
          minTemp: 24.725,
          date: '2021-08-16',
          weatherStateAbbr: 'lr',
          weatherStateImage: 'https://www.metaweather.com/static/img/weather/png/64/lr.png',
          humidity: 68,
          temp: 32.370000000000005,
        },
        {
          maxTemp: 33.825,
          minTemp: 24.685,
          date: '2021-08-17',
          weatherStateAbbr: 'hr',
          weatherStateImage: 'https://www.metaweather.com/static/img/weather/png/64/hr.png',
          humidity: 70,
          temp: 31.630000000000003,
        },
        {
          maxTemp: 33.89,
          minTemp: 25.35,
          date: '2021-08-18',
          weatherStateAbbr: 'hr',
          weatherStateImage: 'https://www.metaweather.com/static/img/weather/png/64/hr.png',
          humidity: 69,
          temp: 31.685000000000002,
        },
        {
          maxTemp: 33.835,
          minTemp: 25.455,
          date: '2021-08-19',
          weatherStateAbbr: 'hr',
          weatherStateImage: 'https://www.metaweather.com/static/img/weather/png/64/hr.png',
          humidity: 70,
          temp: 31.515,
        },
      ],
    });

    return this;
  }

  run() {
    return this.newState;
  }
}

const homeStateFactory = new HomeStateFactory();

export { homeStateFactory };
