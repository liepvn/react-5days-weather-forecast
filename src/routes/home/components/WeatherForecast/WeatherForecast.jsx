import React, { useCallback, useEffect, useState } from 'react';
import { Col, Row, Spinner } from 'react-bootstrap';
import { useIntl } from 'react-intl';
import { useDispatch, useSelector } from 'react-redux';

import { fetchWeatherForecast } from 'routes/home/home.action';

import { ForecastTile } from '../ForecastTile';
import { LocationSearchbox } from '../LocationSearchbox';

import './WeatherForecast.style.scss';

/**
 * Weather forecast component
 *
 * @param {object} selected The selected options
 * @param {function} 
 * @returns {JSX.Element}
 */
export function WeatherForecast({ defaultLocation }) {
  const intl = useIntl();
  const dispatch = useDispatch();
  const [selectedLocation, setSelectedLocation] = useState([defaultLocation]);
  const isLoading = useSelector(({ home }) => home.isLoading);
  const forecast = useSelector(({ home }) => home.forecast);

  useEffect(() => {
    const [location] = selectedLocation;

    location && dispatch(fetchWeatherForecast({ woeid: location.woeid }));
  }, [selectedLocation, dispatch]);

  const handleLocationChange = useCallback((selected) => {
    setSelectedLocation(selected);
  }, []);

  return (
    <div className='weather-forecast'>
      <Row>
        <Col>
          <h2 className='text-center mb-4 text-uppercase text-white'>
            {intl.formatMessage({ id: 'weather.forecast.title', defaultMessage: '5-Days Weather Forecast' })}
          </h2>
        </Col>
      </Row>
      <Row className='mb-2'>
        <Col>
          <LocationSearchbox onChange={handleLocationChange} selected={selectedLocation} />
        </Col>
      </Row>
      <Row>
        <Col>
          {isLoading ? (
            <div className='default-zone border rounded'>
              <Spinner animation="border" variant="info" />
            </div>
          ) : (
            forecast.map((forecast, index) => <ForecastTile key={index} forecast={forecast} />)
          )}
        </Col>
      </Row>
    </div>
  );
}