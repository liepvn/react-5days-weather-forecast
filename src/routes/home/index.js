import Home from './Home';

import homeReducer from './home.reducer';
import homeSaga from './home.saga';

export { homeReducer, homeSaga };

export default Home;
