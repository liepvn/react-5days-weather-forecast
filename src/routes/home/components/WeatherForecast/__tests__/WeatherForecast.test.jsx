import { DEFAULT_LOCATION } from 'routes/home/home.constant';
import { homeState } from 'routes/home/home.reducer';
import { homeStateFactory } from 'routes/home/__fixtures__';
import { render } from 'test/test-utils';

import { WeatherForecast } from '../WeatherForecast';

const mockDispatch = jest.fn();

describe('WeatherForecast', () => {
  it('should render correctly', () => {
    const { getByText, getByPlaceholderText } = render(<WeatherForecast defaultLocation={DEFAULT_LOCATION} />, {
      redux: { mockDispatch, initialState: { home: homeState } },
    });

    expect(getByText('5-Days Weather Forecast')).toBeInTheDocument();
    expect(getByPlaceholderText('Type a city name')).toBeInTheDocument();
  });

  it('should render loading box correctly', () => {
    const { container } = render(<WeatherForecast defaultLocation={DEFAULT_LOCATION} />, {
      redux: { mockDispatch, initialState: { home: homeStateFactory.build({ isLoading: true }).run() } },
    });

    expect(container.getElementsByClassName('default-zone')).toMatchSnapshot();
  });

  it('should render weather forecast tiles correctly', () => {
    const { container } = render(<WeatherForecast defaultLocation={DEFAULT_LOCATION} />, {
      redux: { mockDispatch, initialState: { home: homeStateFactory.build().withSampleForecast().run() } },
    });

    expect(container.getElementsByClassName('forecast-tile')).toMatchSnapshot();
  });
});
