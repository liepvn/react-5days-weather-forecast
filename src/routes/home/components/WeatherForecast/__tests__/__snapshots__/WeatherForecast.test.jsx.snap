// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherForecast should render loading box correctly 1`] = `
HTMLCollection [
  <div
    class="default-zone border rounded"
  >
    <div
      class="spinner-border text-info"
    />
  </div>,
]
`;

exports[`WeatherForecast should render weather forecast tiles correctly 1`] = `
HTMLCollection [
  <div
    class="forecast-tile mb-2 border rounded"
  >
    <div
      class="row"
    >
      <div
        class="col-3"
      >
        <div
          class="d-flex flex-column"
        >
          <p
            class="mb-0 text-uppercase fs-3"
          >
            Sun
          </p>
          <div
            class="humidity"
          >
            <div>
              <i
                class="bi bi-droplet-fill"
              />
               
              <span
                class="text-muted"
              >
                73
                %
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-3"
      >
        <img
          class="w-image"
          src="https://www.metaweather.com/static/img/weather/png/64/hr.png"
        />
      </div>
      <div
        class="d-flex justify-content-end col-6"
      >
        <div
          class="temperature d-flex flex-row"
        >
          <div
            class="temperature__value d-flex flex-column text-end"
          >
            <div>
              <span
                class="fs-3"
              >
                31°C
              </span>
            </div>
            <div
              class="text-muted"
            >
              <span>
                25°C
              </span>
               -
               
              <span>
                32°C
              </span>
            </div>
          </div>
          <div
            class="temperature__icon align-self-center"
          >
            <i
              class="bi bi-thermometer-sun fs-1"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    class="forecast-tile mb-2 border rounded"
  >
    <div
      class="row"
    >
      <div
        class="col-3"
      >
        <div
          class="d-flex flex-column"
        >
          <p
            class="mb-0 text-uppercase fs-3"
          >
            Mon
          </p>
          <div
            class="humidity"
          >
            <div>
              <i
                class="bi bi-droplet-fill"
              />
               
              <span
                class="text-muted"
              >
                68
                %
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-3"
      >
        <img
          class="w-image"
          src="https://www.metaweather.com/static/img/weather/png/64/lr.png"
        />
      </div>
      <div
        class="d-flex justify-content-end col-6"
      >
        <div
          class="temperature d-flex flex-row"
        >
          <div
            class="temperature__value d-flex flex-column text-end"
          >
            <div>
              <span
                class="fs-3"
              >
                32°C
              </span>
            </div>
            <div
              class="text-muted"
            >
              <span>
                25°C
              </span>
               -
               
              <span>
                34°C
              </span>
            </div>
          </div>
          <div
            class="temperature__icon align-self-center"
          >
            <i
              class="bi bi-thermometer-sun fs-1"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    class="forecast-tile mb-2 border rounded"
  >
    <div
      class="row"
    >
      <div
        class="col-3"
      >
        <div
          class="d-flex flex-column"
        >
          <p
            class="mb-0 text-uppercase fs-3"
          >
            Tue
          </p>
          <div
            class="humidity"
          >
            <div>
              <i
                class="bi bi-droplet-fill"
              />
               
              <span
                class="text-muted"
              >
                70
                %
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-3"
      >
        <img
          class="w-image"
          src="https://www.metaweather.com/static/img/weather/png/64/hr.png"
        />
      </div>
      <div
        class="d-flex justify-content-end col-6"
      >
        <div
          class="temperature d-flex flex-row"
        >
          <div
            class="temperature__value d-flex flex-column text-end"
          >
            <div>
              <span
                class="fs-3"
              >
                32°C
              </span>
            </div>
            <div
              class="text-muted"
            >
              <span>
                25°C
              </span>
               -
               
              <span>
                34°C
              </span>
            </div>
          </div>
          <div
            class="temperature__icon align-self-center"
          >
            <i
              class="bi bi-thermometer-sun fs-1"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    class="forecast-tile mb-2 border rounded"
  >
    <div
      class="row"
    >
      <div
        class="col-3"
      >
        <div
          class="d-flex flex-column"
        >
          <p
            class="mb-0 text-uppercase fs-3"
          >
            Wed
          </p>
          <div
            class="humidity"
          >
            <div>
              <i
                class="bi bi-droplet-fill"
              />
               
              <span
                class="text-muted"
              >
                69
                %
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-3"
      >
        <img
          class="w-image"
          src="https://www.metaweather.com/static/img/weather/png/64/hr.png"
        />
      </div>
      <div
        class="d-flex justify-content-end col-6"
      >
        <div
          class="temperature d-flex flex-row"
        >
          <div
            class="temperature__value d-flex flex-column text-end"
          >
            <div>
              <span
                class="fs-3"
              >
                32°C
              </span>
            </div>
            <div
              class="text-muted"
            >
              <span>
                25°C
              </span>
               -
               
              <span>
                34°C
              </span>
            </div>
          </div>
          <div
            class="temperature__icon align-self-center"
          >
            <i
              class="bi bi-thermometer-sun fs-1"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    class="forecast-tile mb-2 border rounded"
  >
    <div
      class="row"
    >
      <div
        class="col-3"
      >
        <div
          class="d-flex flex-column"
        >
          <p
            class="mb-0 text-uppercase fs-3"
          >
            Thu
          </p>
          <div
            class="humidity"
          >
            <div>
              <i
                class="bi bi-droplet-fill"
              />
               
              <span
                class="text-muted"
              >
                70
                %
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-3"
      >
        <img
          class="w-image"
          src="https://www.metaweather.com/static/img/weather/png/64/hr.png"
        />
      </div>
      <div
        class="d-flex justify-content-end col-6"
      >
        <div
          class="temperature d-flex flex-row"
        >
          <div
            class="temperature__value d-flex flex-column text-end"
          >
            <div>
              <span
                class="fs-3"
              >
                32°C
              </span>
            </div>
            <div
              class="text-muted"
            >
              <span>
                25°C
              </span>
               -
               
              <span>
                34°C
              </span>
            </div>
          </div>
          <div
            class="temperature__icon align-self-center"
          >
            <i
              class="bi bi-thermometer-sun fs-1"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
